src/transforms/aggregate.rs

```diff
@@ -0,0 +1,440 @@
+    transforms::{
+        TaskTransform,
+        Transform,
+    },
+    config::{DataType, GlobalOptions, TransformConfig, TransformDescription},
+    event::{
+        metric,
+        Event,
+        EventMetadata,
+    },
+};
+use async_stream::stream;
+use futures::{stream, Stream, StreamExt};
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::HashMap,
+    pin::Pin,
+    time::{Duration},
+};
+
+#[derive(Deserialize, Serialize, Debug, Default, Clone)]
+#[serde(deny_unknown_fields, default)]
+pub struct AggregateConfig {
+    /// The interval between flushes in milliseconds.
+    pub interval_ms: Option<u64>,
+}
+
+inventory::submit! {
+    TransformDescription::new::<AggregateConfig>("aggregate")
+}
+
+impl_generate_config_from_default!(AggregateConfig);
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "aggregate")]
+impl TransformConfig for AggregateConfig {
+    async fn build(&self, _globals: &GlobalOptions) -> crate::Result<Transform> {
+        Aggregate::new(self).map(Transform::task)
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Metric
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Metric
+    }
+
+    fn transform_type(&self) -> &'static str {
+        "aggregate"
+    }
+}
+
+//------------------------------------------------------------------------------
+
+#[derive(Debug)]
+pub struct Aggregate {
+    interval: Duration,
+    map: HashMap<metric::MetricSeries, metric::MetricData>,
+}
+
+impl Aggregate {
+    pub fn new(config: &AggregateConfig) -> crate::Result<Self> {
+        Ok(Self {
+            interval: Duration::from_millis(config.interval_ms.unwrap_or(10 * 1000)),
+            map: HashMap::new(),
+        })
+    }
+
+    fn record(&mut self, event: Event) {
+        let metric = event.as_metric();
+        let series = metric.series();
+        let data = metric.data();
+        let map = &mut self.map;
+
+        match data.kind {
+            metric::MetricKind::Incremental => {
+                match map.get_mut(&series) {
+                    // We already have something, add to it, will update timestamp as well.
+                    Some(existing) => existing.update(data),
+                    None => {
+                        // New so store
+                        map.insert(series.clone(), data.clone());
+                        true
+                    }
+                };
+            },
+            metric::MetricKind::Absolute => {
+                // Always replace/store
+                map.insert(series.clone(), data.clone());
+            }
+        };
+
+        emit!(AggregateEventRecorded);
+    }
+
+    fn flush_into(&mut self, output: &mut Vec<Event>) -> u64 {
+        let mut count = 0_u64;
+        let map = &mut self.map;
+
+        if map.len() > 0 {
+            // TODO: not clear how this should work with aggregation so just stuffing a default one
+            // in for now.
```

Comment:
Good point, I'll have to look into this.

Reply:
Done & TODO removed. See https://github.com/timberio/vector/pull/7846/files#r653754356 for a hiccup I ran into though.